File: game_of_life.cpp
// Jeu de la vie en C++
#include <iostream>
#include <vector>
#include <chrono>
#include <thread>
#include <cstdlib>
#include <ctime>

using Grid = std::vector<std::vector<int>>;

int countNeighbors(const Grid& grid, int x, int y) {
    int count = 0;
    int rows = grid.size();
    int cols = grid[0].size();
    for (int i = -1; i <= 1; ++i) {
        for (int j = -1; j <= 1; ++j) {
            if (i == 0 && j == 0) continue;
            int nx = x + i;
            int ny = y + j;
            if (nx >= 0 && nx < rows && ny >= 0 && ny < cols) {
                count += grid[nx][ny];
            }
        }
    }
    return count;
}

void printGrid(const Grid& grid) {
    std::cout << "\033[2J\033[H"; // efface l'écran et remonte le curseur
    for (const auto& row : grid) {
        for (int cell : row) {
            std::cout << (cell ? 'O' : ' ');
        }
        std::cout << '\n';
    }
}

int main() {
    const int rows = 20;
    const int cols = 40;
    Grid grid(rows, std::vector<int>(cols));

    // Initialisation aléatoire
    std::srand(static_cast<unsigned>(std::time(nullptr)));
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            grid[i][j] = std::rand() % 2;
        }
    }

    while (true) {
        printGrid(grid);
        Grid next = grid;
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                int alive = countNeighbors(grid, i, j);
                if (grid[i][j]) {
                    next[i][j] = (alive == 2 || alive == 3) ? 1 : 0;
                } else {
                    next[i][j] = (alive == 3) ? 1 : 0;
                }
            }
        }
        grid.swap(next);
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
    }

    return 0;
}

File: game_of_life.py
#!/usr/bin/env python3
# Jeu de la Vie en Python

import random
import time
import os
import sys

Grid = list[list[int]]

def count_neighbors(grid: Grid, x: int, y: int) -> int:
    rows, cols = len(grid), len(grid[0])
    count = 0
    for i in (-1, 0, 1):
        for j in (-1, 0, 1):
            if i == 0 and j == 0:
                continue
            nx, ny = x + i, y + j
            if 0 <= nx < rows and 0 <= ny < cols:
                count += grid[nx][ny]
    return count


def print_grid(grid: Grid) -> None:
    sys.stdout.write("\033[2J\033[H")  # Efface l'écran et replace le curseur en haut
    for row in grid:
        for cell in row:
            sys.stdout.write('O' if cell else ' ')
        sys.stdout.write('\n')
    sys.stdout.flush()


def main() -> None:
    rows, cols = 20, 40
    # Initialisation aléatoire
    grid: Grid = [[random.randint(0, 1) for _ in range(cols)] for _ in range(rows)]

    try:
        while True:
            print_grid(grid)
            next_grid: Grid = [[0] * cols for _ in range(rows)]
            for i in range(rows):
                for j in range(cols):
                    alive = count_neighbors(grid, i, j)
                    if grid[i][j]:
                        next_grid[i][j] = 1 if alive in (2, 3) else 0
                    else:
                        next_grid[i][j] = 1 if alive == 3 else 0
            grid = next_grid
            time.sleep(0.2)
    except KeyboardInterrupt:
        print("\nArrêt du programme.")


if __name__ == "__main__":
    main()